src/preprocessing.ml

```diff
@@ -477,9 +477,70 @@ let get_compat_ppx_exe sctx ~name ~kind =
     in
     ppx_exe sctx ~key ~dir_kind:Jbuild
 
-let get_ppx_driver sctx ~loc ~scope ~dir_kind pps =
+let args_of_cookies cookies =
+  List.map ~f:(fun (name, value) ->
+    ["--cookie"; sprintf "%s=%S" name value]
+  ) cookies |> List.flatten
+
+let library_name_cookie : Lib_kind.Ppx_args.Cookie.t =
+  {
+    name = "library_name";
+    value = String_with_vars.make_var Loc.none "library_name"
+  }
+let get_cookies ~loc ~expander ~lib_name libs =
+  let expander =
+    match lib_name with
+    | None -> expander
+    | Some name ->
+      let bindings =
+        Pform.Map.singleton "library_name"
+          (Values [String (Lib_name.Local.to_string name)])
+      in
+      Expander.add_bindings expander ~bindings
+  in
+  try
+    Ok (libs
+      |> List.map ~f:
+        (fun t ->
+          match Lib.kind t with
+          | Normal -> []
+          | Ppx_rewriter {cookies}
+          | Ppx_deriver {cookies} ->
+            List.map ~f:(fun {Lib_kind.Ppx_args.Cookie.name; value} ->
+              (name, (Expander.expand_str expander value, Lib.name t)))
+              (if lib_name = None then cookies else library_name_cookie :: cookies)
+        )
+      |> List.flatten
+      |> String.Map.of_list_reducei ~f:
+        (fun name ((val1, lib1) as res) (val2, lib2) ->
+          if String.equal val1 val2 then
+            res
+          else
+            let lib1 = Lib_name.to_string lib1 in
+            let lib2 = Lib_name.to_string lib2 in
+            Errors.fail loc "%a" Fmt.text
+             (sprintf "%s and %s have inconsistent requests for cookie \"%s\"; \
+                        %s requests \"%s\" and %s requests \"%s\""
```

Comment:
I would use `%S` instead of `"%s"` in this format string. This way, if the name or value contain `"` or non-printable characters, they will be escaped.